pkg/auth/token_handler.go

```diff
@@ -0,0 +1,259 @@
+	}
+
+	return tokenRequestResponse.Status.Token, nil
+}
+
+// VerifyJWT verifies the given JWT for the expected audience and returns the parsed ID token.
+func (c *OIDCTokenHandler) VerifyJWT(ctx context.Context, jwt, audience string) (*IDToken, error) {
+	c.providerLock.RLock()
+	defer c.providerLock.RUnlock()
+
+	if c.provider == nil {
+		return nil, fmt.Errorf("provider is nil. Is the OIDC provider config correct?") // for devs this could also be a sign, that OIDCTokenHandler.Start() wasn't called
+	}
+
+	verifier := c.provider.Verifier(&oidc.Config{
+		ClientID: audience,
+	})
+
+	token, err := verifier.Verify(ctx, jwt)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify JWT: %w", err)
+	}
+
+	return &IDToken{
+		Issuer:          token.Issuer,
+		Audience:        token.Audience,
+		Subject:         token.Subject,
+		Expiry:          token.Expiry,
+		IssuedAt:        token.IssuedAt,
+		AccessTokenHash: token.AccessTokenHash,
+	}, nil
+}
+
+func (c *OIDCTokenHandler) syncProviderConfigPeriodically(stopCh <-chan struct{}) {
+	ticker := time.NewTicker(c.providerSyncInterval)
+
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				if err := c.updateProviderConfig(); err != nil {
+					c.logger.Error("could not update OIDC provider config.", zap.Error(err))
+				}
+			case <-stopCh:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (c *OIDCTokenHandler) updateProviderConfig() error {
+	featureFlags := c.featureStore.Load()
+	if !featureFlags.IsOIDCAuthentication() {
+		// we can skip if OIDC is not enabled
+		c.logger.Debug("oidc authentication is disabled, so skipping to update provider config")
+		return nil
+	}
+
+	c.logger.Debug("updating OIDC provider config...")
+
+	ctx := context.Background()
+
+	// load OIDC configuration from API server
+	oidcConfig, err := c.loadKubernetesOIDCConfiguration(ctx)
+	if err != nil {
+		return fmt.Errorf("could not get clusters OIDC configuration: %w", err)
+	}
+
+	if strings.HasPrefix(oidcConfig.Issuer, "https://kubernetes.default.svc") {
+		// setup http client for OIDC lib with the CA of the API server
+		httpClient, err := c.getHTTPClientForKubeAPIServer()
+		if err != nil {
+			return fmt.Errorf("could not get HTTP client with TLS certs of API server: %w", err)
+		}
+		ctx = oidc.ClientContext(context.Background(), httpClient)
+	} else {
+		// TODO: if OIDC issuer is an external one, configure http client with certs for external issuer
+		// (e.g. load certs from a config map)
+	}
+
+	// get OIDC provider
+	provider, err := oidc.NewProvider(ctx, oidcConfig.Issuer)
+	if err != nil {
+		return fmt.Errorf("could not get OIDC provider: %w", err)
+	}
+
+	c.providerLock.Lock()
+	defer c.providerLock.Unlock()
+
+	c.provider = provider
+
+	c.logger.Debug("updated OIDC provider config", zap.Any("oidc-config", oidcConfig))
+
+	return nil
+}
+
+func (c *OIDCTokenHandler) getHTTPClientForKubeAPIServer() (*http.Client, error) {
+	cfg, err := rest.InClusterConfig()
```

Comment:
knative/pkg uses a more advanced way of resolving REST client config [1] which enabled you to configure it through flags, etc, can we reuse it here?

https://github.com/knative/pkg/blob/6cf4b051de4f9859b0f1b80f790a83e84a75bc5b/environment/client_config.go

Reply:
To get the restConfig, I would have to do something like the following IIUC:
```
env := &environment.ClientConfig{}
env.InitFlags(flag.CommandLine)
flag.Parse()
restConfig, err := env.GetRESTConfig()
```
But the same flags were added before already, because `EnableInjectionOrDie()` might call it too (via `ParseAndGetRESTConfigOrDie()`): https://github.com/knative/eventing/blob/3dfc2ea9565e54b0146011bc9784120c3fd004f5/vendor/knative.dev/pkg/injection/injection.go#L43-L68

And then this restConfig is available in the application context (line 58). That would mean, that we would need pass it always to `NewOIDCTokenHandler` (and add it in all the Reconciler structs, which use the oidcTokenManager later) :shrug: 
WDYT?

Reply:
Can't we do `injection.GetConfig(ctx)` in the `NewOIDCTokenHandler`?

Reply:
yes. We only need to make sure we always pass an application context (which should be fine)

Reply:
And IIUC update the `reconciler/testing.SetupFakeContext()` to add a config to the context (IIUC), as this is used by some tests (and thus the caller of `NewOIDCTokenHandler()` does not need to know we need a restconfig in the context)